sdk/src/OpenTelemetry/Processor/Batch.hs

```diff
@@ -237,27 +268,66 @@ batchProcessor BatchTimeoutConfig {..} exporter = liftIO $ do
+          when appendFailed $ void $ atomically $ tryPutTMVar workSignal ()
+      , processorForceFlush = void $ atomically $ tryPutTMVar workSignal ()
       , -- TODO where to call restore, if anywhere?
-        processorShutdown = async $ mask $ \_restore -> do
-          -- flush remaining messages
-          void $ tryPutMVar workSignal Shutdown
-
-          shutdownResult <-
-            timeout (millisToMicros exportTimeoutMillis) $
-              wait worker
-          -- make sure the worker comes down
-          uninterruptibleCancel worker
-          -- TODO, not convinced we should shut down processor here
-
-          case shutdownResult of
-            Nothing -> pure ShutdownFailure
-            Just _ -> pure ShutdownSuccess
+        processorShutdown =
+          asyncWithUnmask $ \unmask -> unmask $ do
+            -- we call asyncWithUnmask here because the shutdown action is
+            -- likely to happen inside of a `finally` or `bracket`. the
+            -- @safe-exceptions@ pattern (followed by unliftio as well)
+            -- will use uninterruptibleMask in an exception cleanup. the
+            -- uninterruptibleMask state means that the `timeout` call
+            -- below will never exit, because `wait worker` will be in the
+            -- `uninterruptibleMasked` state, and the timeout async
+            -- exception will not be delivered.
+            --
+            -- see note [Unmasking Asyncs]
+            mask $ \_restore -> do
+              -- is it a little silly that we unmask and remask? seems
+              -- silly! but the `mask` here is doing an interruptible mask.
+              -- which means that async exceptions can still be delivered
+              -- if a process is blocking.
+
+              -- flush remaining messages and signal the worker to shutdown
+              void $ atomically $ putTMVar shutdownSignal ()
+
+              -- gracefully wait for the worker to stop. we may be in
+              -- a `bracket` or responding to an async exception, so we
+              -- must be very careful not to wait too long. the following
+              -- STM action will block, so we'll be susceptible to an async
+              -- exception.
+              delay <- registerDelay (millisToMicros exportTimeoutMillis)
+              shutdownResult <-
+                atomically $
+                  msum
+                    [ Just <$> waitCatchSTM worker
+                    , const Nothing <$> do
+                        shouldStop <- readTVar delay
+                        check shouldStop
+                    ]
+
+              -- make sure the worker comes down. does not wait for the
+              -- worker thread to complete. wraps in SomeAsyncException to
+              -- ensure that any exception handler in the worker know sit
+              -- is an async exception.
+              throwTo (asyncThreadId worker) (SomeAsyncException AsyncCancelled)
```

Comment:
Previously this was `uninterruptibleCancel` - `uninterruptibleMask_ . cancel`. And `cancel (Async tid wait) = atomically wait <* throwTo tid AsyncCancelled`. So.... `cancel` actually *blocks* until the async thread has terminated.

With `uninterruptibleCancel`, if the `Async` is *stuck*, then we're also *stuck*, and the only way to escape is with the RTS killing everything abruptly.

Rather than kill everything abruptly, let's just *not block forever*, and hope that the worker isn't irreparably stuck.